### PR TITLE
Support non-conforming meshes with BlockNonlinearForm and ParBlockNonlinearForm

### PR DESCRIPTION
**Pull request by jandrej**
This PR adds support for support non-conforming meshes with BlockNonlinearForm and ParBlockNonlinearForm. I made changes to ex19 and ex19p to provide test cases.

I changed the name and return type of an internal function in BlockNonlinearForm to reflect the functionality change.

The changes work in a dependent code as well.

Obligatory GLVis snapshot
Screen Shot 2020-09-07 at 11 58 41 AM

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1744](https://github.com/mfem/mfem/pull/1744) | @jandrej | @tzanio | @jakubcerveny + @jamiebramwell + @mlstowell | 09/08/20 | 09/22/20 | 10/04/20 | |

### CODE REVIEW

**Comment by jandrej** [comment]
LGTM

**Comment by jandrej** [comment]
LGTM

File: fem/nonlinearform.hpp [hunk 1]
```diff
@@ -199,24 +199,41 @@ class BlockNonlinearForm : public Operator
        GridFunction-like block-vector data (e.g. in parallel). */
    mutable BlockVector xs, ys;
 
-   mutable Array2D<SparseMatrix*> Grads;
+   mutable Array2D<SparseMatrix*> Grads, cGrads;
```
Comment (jamiebramwell) [inline fem/nonlinearform.hpp:202]: what does `cGrads` stand for? Conforming grads? I think a more verbose name would be useful.

File: fem/nonlinearform.hpp [hunk 2]
```diff
@@ -199,24 +199,41 @@ class BlockNonlinearForm : public Operator
...
-   // Essential vdofs: one list of vdofs for each space in 'fes'
-   Array<Array<int> *> ess_vdofs;
+   // Array of Arrays of tdofs for each space in 'fes'
+   Array<Array<int> *> ess_tdofs;
+
+   /// Array of pointers to the prolongation matrix of fes, may be NULL
+   Array<const Operator *> P;
+
+   /// Array of results of dynamic-casting P to SparseMatrix pointer
+   Array<const SparseMatrix *> cP;
```
Comment (jamiebramwell) [inline fem/nonlinearform.hpp:216]: I assume this is cast prolongation. I longer name here would also be useful.

**Review by jamiebramwell — APPROVED** [review]
Thank you for figuring this out! I just have some minor naming quibbles, but otherwise it looks great!

**Review by jakubcerveny — APPROVED** [review]
I don't really know how this works (and the PR doesn't describe the changes for the non-expert), but I tested a couple of runs, ran with valgrind, ran a 3D case with elements refined towards a vertex, and all seems to work...

beam-hex

**Review by mlstowell — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Review by tzanio — APPROVED** [review]
Thanks for the fix @jandrej!

@jakubcerveny, @jamiebramwell, @mlstowell and @v-dobrev -- Thanks for the review and fixes!

**Pull request merged**